docs/performance/inserts/architecture.txt

```diff
@@ -0,0 +1,250 @@
+important that you have a basic understanding of how CrateDB handles inserts
+internally. This will make it easier to reason about the location of potential
+performance bottlenecks.
+
+.. rubric:: Table of Contents
+
+.. contents::
+   :local:
+
+Types of Insert
+===============
+
+There are three kinds of inserts as far as CrateDB's insert handling is
+concerned.
+
+Single Inserts
+--------------
+
+A single insert looks like this:
+
+.. code-block:: psql
+
+  cr> INSERT INTO my_table (id, first_name)
+  ...      VALUES (1, "Arthur");
+  INSERT OK, 1 row affected  (... sec)
+
+As far as CrateDB's insert handling is concerned,
+:ref:`client batching <client_feature_method>` (e.g. JDBC's
+``addBatch`` and ``executeBatch``) counts as a type one insert, because client
+batching is little more than an efficient way of issuing a collection of single
+inserts.
+
+Inserts With Multiple Value Expressions
+---------------------------------------
+
+An insert with :ref:`multiple value expressions <multiple_value_expressions>`
+looks like this:
+
+.. code-block:: psql
+
+   cr> INSERT INTO my_table (id, first_name)
+   ...      VALUES (1, 'Arthur'),
+   ...             (2, 'Trillian'),
+   ...             (2, 'Marvin');
+   INSERT OK, 3 rows affected  (... sec)
+
+Bulk Inserts
+------------
+
+Bulk inserts are typically done via :ref:`the CrateDB HTTP endpoint
+<switcing_to_bulk_inserts>`, but they can also be done with SQL (see
+:ref:`unnest_method`) if the HTTP endpoint isn't a viable option.
+
+The ``UNNEST`` method looks like this:
+
+  .. code-block:: psql
+
+     cr> INSERT INTO my_table (id, first_name)
+     ...   (SELECT *
+     ...      FROM UNNEST(
+     ...             [1, 2, 3],
+     ...             ['Arthur', 'Trillian', 'Marvin']));
+     INSERT OK, 3 rows affected  (... sec)
+
+Processing Phases
+=================
+
+Parsing and Analysing
+---------------------
+
+When a user issues an insert statement to the cluster then this request is sent
+to whichever individual node the client connected to. This node is then
+responsible for performing statement parsing and analysing.
+
+The statement is parsed by the *SQL parser*. The *parsed statement* is then
+analysed by the *statement analyser*, which produces an *analysed statement*.
+
+The statement analyser performs the following checks:
+
+- Does the target table exist?
+
+- Do the specified table columns exist?
+
+   - If not, does the table have a `dynamic column policy`_?
+
+- Do the data types of the insert values match the data types of the existing columns?
+
+- Are all ``PRIMARY KEY`` columns and ``NOT NULL`` column specified?
+
+- Are the values for ``NOT NULL`` columns non null?
+
+- Is the table partitioned?
+
+  - If so, does a `partition`_ exist for all insert values?
+
+    - If not, enrich the statement with a list of the partitions that need to be created
+
+Preparation
+-----------
+
+TODO: could this also be called the "planning" phase?
```

Comment:
No, Not exactly.  For inserts the planner does a trivial job to create an InsertPlan from the InsertAnalyzedStatement. The Preparation described here is part of the Execution, like a pre-Execution step. So the component from now one is the Executor.